airbyte-integrations/bases/base-python/base_python/sdk/streams/core.py

```diff
@@ -0,0 +1,104 @@
+import inspect
+from abc import ABC, abstractmethod
+from typing import Mapping, Any, Iterable, List, Union
+
+from airbyte_protocol import AirbyteStream, SyncMode
+from base_python.logger import AirbyteLogger
+from base_python.schema_helpers import ResourceSchemaLoader
+
+
+def package_name_from_class(cls: object) -> str:
+    """Find the package name given a class name"""
+    module = inspect.getmodule(cls)
+    return module.__name__.split(".")[0]
+
+class Stream(ABC):
+    # Use self.logger in subclasses to log any messages
+    logger = AirbyteLogger()  # TODO use native "logging" loggers with custom handlers
+
+    @property
+    def name(self) -> str:
+        """
+        :return: Stream name. By default this is the implementing class name, but it can be overridden as needed.
+        """
+        return self.__class__.__name__
+
+    @abstractmethod
+    def read_stream(self, stream_state: Mapping[str, Any] = None) -> Iterable[Mapping[str, Any]]:
+        """
+        This method should be overridden by subclasses
+
+        :param stream_state: State dict to use when extracting records.
+        :return: A generator which yields all the records in this stream. Each record is a dict from properties to their values matching the schema
+         of the stream.
+        """
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        """
+        :return: A dict of the JSON schema representing this stream.
+
+        The default implementation of this method looks for a JSONSchema file with the same name as this stream's "name" property.
+        Override as needed.
+        """
+        # TODO show an example of using pydantic to define the JSON schema, or reading an OpenAPI spec
+        # TODO change to snakecase by default
+        return ResourceSchemaLoader(package_name_from_class(self.__class__)).get_schema(self.name)
+
+    def as_airbyte_stream(self) -> AirbyteStream:
+        return AirbyteStream(
+            name=self.name,
+            json_schema=dict(self.get_json_schema()),
+            supported_sync_modes=[SyncMode.full_refresh]
+        )
+
+class IncrementalStream(Stream, ABC):
+    @property
+    def cursor_field(self) -> Union[str, List[str]]:
+        """
+        Override to return the name of the default cursor field used by this stream e.g: an API entity might always use created_at as the cursor field.
```

Comment:
should this method be called `default_cursor_field`? it sounds like this is only populated if the source is setting the cursor field if i'm reading the comment correctly. or maybe call this method `source_defined_cursor` and then the method be low is just `has_source_defined_cursor` and default impl can just check if this method returns null / empty array

Reply:
I think there is either cursor_field or not, what makes it default_cursor_field?
also what is the case when we need additional property `source_defined_cursor`?
isn't `cursor_field` self sufficient? i.e. it is either null/[] or str/[str]

Reply:
`cursor_field` in the general case could be chosen by a user. For example when replicating data from a SQL database, you could choose to replicate based on `created_at` or `updated_at`. In this case, because the cursor is configurable by the user, `source_defined_cursor` would return False even though the cursor field is defined.

Reply:
1. I think base abstraction should include this logic of setting custom cursor field
2. I think the name `source_defined_cursor` could be confusing, maybe something like `cursor_field_overrided`?
3. If we have `cursor_field` and `default_cursor_field`, wouldn't `source_defined_cursor` implementation just compare if one is equal to another? Also I still don't see the case when the stream might need this information.

Reply:
1. Added this
2. Agreed with you. Unfortunately currently this is at the protocol level so changing this is a heavy lift because it would change all connectors, the UI, and the core API. To make this better I defaulted all HTTP streams to `True` on this field so the user doesn't need to think about this (most HTTP sources have a static cursor field anyways unlike e.g: databases)